test/unit/mem/view/src/Accessor.cpp

```diff
@@ -0,0 +1,130 @@
+/* Copyright 2021 Bernhard Manfred Gruber
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <alpaka/example/ExampleDefaultAcc.hpp>
+#include <alpaka/mem/view/Accessor.hpp>
+
+#include <catch2/catch.hpp>
+
+namespace
+{
+    constexpr auto N = 1024;
+
+    struct WriteKernelTemplate
+    {
+        template<typename Acc, typename Accessor>
+        ALPAKA_FN_ACC void operator()(Acc const& acc, Accessor data) const
+        {
+            (void) acc;
+            data[1] = 1.0f;
+            data(2) = 2.0f;
+            data[alpaka::Vec<alpaka::Dim<Acc>, alpaka::Idx<Acc>>{alpaka::Idx<Acc>{3}}] = 3.0f;
+        }
+    };
+
+    struct WriteKernelExplicit
+    {
+        template<typename Acc, typename Pointer, typename Idx, typename Dim>
+        ALPAKA_FN_ACC void operator()(Acc const& acc, alpaka::Accessor<Pointer, Idx, Dim> data) const
+        {
+            (void) acc;
+            data[1] = 1.0f;
+            data(2) = 2.0f;
+            data[alpaka::Vec<alpaka::Dim<Acc>, alpaka::Idx<Acc>>{alpaka::Idx<Acc>{3}}] = 3.0f;
+        }
+    };
+
+    struct ReadKernelTemplate
+    {
+        template<typename Acc, typename Accessor>
+        ALPAKA_FN_ACC void operator()(Acc const& acc, Accessor data) const
```

Comment:
Here, the only way to communicate, that this is a read-only accessor, is to either make it const like `const Accessor data` or have a real C++20 concept like `Accessor<read_only> auto data`.

We should also consider, if we need to add more access modes, like `write_only`. Or some kind of uncached access.

Reply:
For the record, SYCL defines the following interface: `accessor</* other stuff */, access_mode::{read, write, read_write}>`.

Reply:
The SYCL way makes perfect sense if there are more types of accesses than read only and read/write. Otherwise we could just leverage const.

Reply:
Write only makes sense for FPGAs. With this knowledge memory accesses can be accelerated because the write operation has more memory ports available.

Reply:
We can certainly go down this road, but it is yet another template parameter on `Accessor`.

Reply:
In redGraps we define also an acces mode atomic. It is like write but give us the possibility to express that multiple kernel/taks can run in parallel even if the manipulate the data.

Reply:
Well, this raises the question if you would want to use accessors for scheduling similar to SYCL, which so far I do not forsee. But you could actually get this information in `alpaka::exec()` because you know the types of the accessors passed in.

Btw, since you mentined atomic access. I would rather prefer a solution similar to `std::atomic_ref<T>` than declaring an accessor atomic.

Reply:
> if you would want to use accessors for scheduling similar to SYCL

No, I think we should leave that to redGrapes. Although this will probably cause some internal synchronization in the SYCL backend because we can basically map SYCL accessors 1:1 to alpaka's accessors.

> I would rather prefer a solution similar to std::atomic_ref<T> than declaring an accessor atomic.

That is the way SYCL went, too. In SYCL 2020 they abandoned old-style atomics and basically just cloned `std::atomic_ref`.

Reply:
> Btw, since you mentined atomic access. I would rather prefer a solution similar to std::atomic_ref<T> than declaring an accessor atomic.

I do not mean providing an accessor that is atomic, IMO an accessor can have multiple access policies instead of a single. I know that this is more complex in the design but useful to optimize the scheduling of tasks.
The multi-policy access types can be avoided if you add an access mode priority list similar to the priority for C++ operators.

Reply:
@psychocoderHPC I am not getting what use case we are trying to solve with this access policies. Could you please provide an example to me? Let's say we have a kernel that takes a 2D buffer of a type `Vector` like this:

```c++
struct Vector {
    float x, y, z;
};

struct Kernel {
    template <....>
    void operator()(auto& acc, alpaka::Accessor<Vector*, Vector, Dim, Idx> buffer) {
        // what do you want to do here? Except something like:
        Vector v = buffer(x, y);
        buffer(x, y) = v;
    }
};
```

How does an access policy change reading a `Vector` from the location `x, y`. What do you need to customize?

Reply:
From an accessor, you can get the information on how you access the memory e.g. `read`. If you have two kernels where you read the same data then you know you can enqueue both kernels so that they can run in parallel. 
If you have kernel A with an accessor to a buffer X(read) and a kernel B with an accessor x(write) you not to enqueue it sequential in the given order to guarantee that you have no races.
If you have a kernel A with X(atomic) and B with X(atomic) you can also enqueue both parallels. But a kernel C with X(write) should be enqueued after A and B.
This is IMO the reason why we need more than read and write as access policy.

Why do we need more than one access policy?

If you give an access X(write) to the device this does not always mean you would like to write only to the buffer. Maybe you **read** from this buffer first e.g. to cache data. Within the device, you should IMO have a way to access the memory. 
Let's call an iterator with random access cursor.
This could be `X(write)->getCursor(read)` here you have now the possibility to optimize based on the information the user is giving to `getCursor`. Remember I set only the policy `write`, therefore `X(write)->getCursor(read)` should fail to avoid that a user can query different access than he configured during the kernel call. 
To be valid a kernel should started with `X(write,read)` which than allows the user on the device to query `X(write,read)->getCursor(read)` or `X(write,read)->getCursor(write)`.

As I saw in your draft you do not have the `getCursor()` stage in your concepts but I think they are mandatory for optimization and to validate that the user is only performing access on the device which he applied for on host side.